src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs

```diff
@@ -0,0 +1,464 @@
+                    .ToDictionary(pair => pair.Key, pair => pair.Value, StringComparer.OrdinalIgnoreCase);
+
+            }
+            else
+            {
+                parameters = CoreHelpers.SplitWithQuotes(wwwAuthenticateValue, ',')
+                    .Select(v => AuthenticationHeaderParser.ExtractKeyValuePair(v.Trim()))
+                    .ToDictionary(pair => pair.Key, pair => pair.Value, StringComparer.OrdinalIgnoreCase);
+            }
+
+            return CreateWwwAuthenticateParameters(parameters);
+        }
+
+        internal static WwwAuthenticateParameters CreateWwwAuthenticateParameters(IDictionary<string, string> values)
+        {
+            WwwAuthenticateParameters wwwAuthenticateParameters = new WwwAuthenticateParameters
+            {
+                RawParameters = values
+            };
+
+            if (values.Count == 0)
+            {
+                //unable to parse auth header values
+                wwwAuthenticateParameters.ParsingError = true;
+            }
+
+            string value;
+
+            if (values.TryGetValue("authorization_uri", out value))
+            {
+                wwwAuthenticateParameters.Authority = value.Replace("/oauth2/authorize", string.Empty);
+            }
+
+            if (string.IsNullOrEmpty(wwwAuthenticateParameters.Authority))
+            {
+                if (values.TryGetValue("authorization", out value))
+                {
+                    wwwAuthenticateParameters.Authority = value.Replace("/oauth2/authorize", string.Empty);
+                }
+            }
+
+            if (string.IsNullOrEmpty(wwwAuthenticateParameters.Authority))
+            {
+                if (values.TryGetValue("authority", out value))
+                {
+                    wwwAuthenticateParameters.Authority = value.TrimEnd('/');
+                }
+            }
+
+            if (values.TryGetValue("claims", out value))
+            {
+                wwwAuthenticateParameters.Claims = GetJsonFragment(value);
+            }
+
+            if (values.TryGetValue("error", out value))
+            {
+                wwwAuthenticateParameters.Error = value;
+            }
+
+            if (values.TryGetValue("nonce", out value))
+            {
+                wwwAuthenticateParameters.ServerNonce = value.Replace("\"", string.Empty);
+            }
+
+            return wwwAuthenticateParameters;
+        }
+
+        /// <summary>
+        /// Checks if input is a base-64 encoded string.
+        /// If it is one, decodes it to get a json fragment.
+        /// </summary>
+        /// <param name="inputString">Input string</param>
+        /// <returns>a json fragment (original input string or decoded from base64 encoded).</returns>
+        private static string GetJsonFragment(string inputString)
+        {
+            if (string.IsNullOrEmpty(inputString) || inputString.Length % 4 != 0 || inputString.Any(c => char.IsWhiteSpace(c)))
```

Comment:
I am not sure about this. How about: 

1. Try to base64url decode it. If this works and the result is JSON, use it.
2. Try to see if the result is JSON. If it is, use it.
3. Fail with parsing error?

Reply:
Yeah, I didnt write this so not 100% sure what the checks are looking for. ill see if I can rework it.